packages/engine-core/src/data-proxy/DataProxyEngine.ts

```diff
@@ -153,76 +167,119 @@ export class DataProxyEngine extends Engine {
       isolationLevel: transaction?.isolationLevel,
     }
 
-    const { batchResult } = await this.requestInternal<T>(body, headers, attempt)
+    const { batchResult } = await this.requestInternal<T, true>(body, headers)
 
+    // TODO: add elapsed to each result similar to BinaryEngine
+    // also check that the error handling is correct for batch
     return batchResult
   }
 
-  private async requestInternal<T>(body: Record<string, any>, headers: Record<string, string>, attempt: number) {
-    try {
-      this.logEmitter.emit('info', {
-        message: `Calling ${await this.url('graphql')} (n=${attempt})`,
-      })
+  private requestInternal<T, Batch extends boolean = false>(
+    body: Record<string, any>,
+    headers: QueryEngineRequestHeaders,
+    itx?: InteractiveTransactionOptions<DataProxyTxInfoPayload>,
+  ): Promise<Batch extends true ? { batchResult: QueryEngineResult<T>[] } : QueryEngineResult<T>> {
+    return this.withRetry({
+      actionGerund: 'querying',
+      callback: async ({ logHttpCall }) => {
+        const url = itx ? `${itx.payload.endpoint}/graphql` : await this.url('graphql')
+
+        logHttpCall(url)
+
+        const response = await request(url, {
+          method: 'POST',
+          headers: { ...runtimeHeadersToHttpHeaders(headers), ...this.headers },
```

Comment:
```suggestion
          headers: runtimeHeadersToHttpHeaders(headers)
```

I think `runtimeHeadersToHttpHeaders` preserves all original headers, except `transactionId`?

Reply:
`headers` comes from the function argument and isn't a superset of `this.headers` (which is basically just the `authorization` header), so we need both.

Technically we shouldn't even need `runtimeHeadersToHttpHeaders(headers)` here anymore with the recent Data Proxy changes and could just make it `headers: this.headers` and move `runtimeHeadersToHttpHeaders` back to `BinaryEngine` as a private function, but let's better do that after the release probably. There's no harm in sending those headers, but we haven't tested not sending them yet and I'd avoid doing that change last minute. And they would have to come back when we support tracing with Data Proxy anyway (although we could filter out `transactionId` specifically).